### DISCUSSION TITLE
Project Potential

### DISCUSSION THREAD

**Discussion by AdmiralBongo** [dsharlet/LiveSPICE#221] [Ideas]
Disclaimer: I know there will be open-source/freeware idealists and whatnot and I sympathize, but just hear me out.

Ever since I stumbled upon LiveSPICE, I've found it to hold a huge potential to be something outright revolutionary. Not as a tool for merely trying out various designs for boutique pedal builders in a more convenient way, but as a tool for producers and guitar/bass players, both live and in studio, allowing to simulate literally any amp and any pedal, fully tweakable, including custom designs that have never existed in real life, allowing to share them as people do with Neural Amp Modeler. Problem is how raw and poorly optimized it is at the moment. Thus, I've got a question: have you considered seeking out investors and monetizing the project in order to hire some more programmers and turn LiveSPICE into something bigger?

**Comment by dsharlet** [comment]
Thank you for the interesting question, and I'm honored by the proposition :) I think both free/open source and commercial software both have their place, that's not why I haven't tried to do more with LiveSPICE.

The big blocker is that I think there are some major open research questions:

- LiveSPICE circuits are quite limited in complexity, before performance becomes an issue. More importantly, it's kind of hard to predict and understand where the limits are. I don't think this is just a matter of more engineering optimization, it's really an algorithm problem. You might be able to make LiveSPICE 2x faster with a lot more engineering, but I think an algorithmic speedup is necessary to be a viable product, not just constant factor speedups.
- Not being able to support truly dynamic potentiometers/parameters is a really major limitation that I think limits LiveSPICE to more of a "free hobby level tool" than a really practical tool. @Federerer was making some progress on this, but I think even this approach might only scale to 1 or 2 dynamic parameters (TBD though). 

I don't know that funding or more people will help with these problems. They are more like math problems than software problems. Hiring someone to solve these problems would be very difficult, it's a very narrow expertise. I work with a bunch of people like this (in my regular work), and we've talked about how to fix these problems, we've never come up with an idea for how to make real dents in these problems. These kinds of problems are more of a PhD thesis project (or a few) than a startup. I wouldn't want to accept funding for hiring people without a clear line of sight to making a lot of progress on these issues.

I think the only real way forward on both issues is to fully commit to a much more approximate modeling approach (Neural Amp Modeler being an extreme example). But, I think what makes LiveSPICE interesting today is that it is a real physical model and simulation, like other SPICE software :) And setting that aside, building a good approximate model that can be trusted to accurately represent a circuit, while supporting dynamic parameters, might be an even harder problem than solving the two I mentioned above!

**Reply by AdmiralBongo** [reply]
Thanks for a reply! It's a bit sad that there are so many complications about it. But then again, just consider this idea. From what I've heard, Line 6 has a hybrid approach to the performance issue: they use neural modeling of separate components in Helix and combine the captures into full-fledged circuits. Plus, NAM is open-source and apparently its developer is open to allowing commercial plugin developers borrowing its source code. Given that a large enough library of neurally-modeled components is built, I think the performance may be improved. That being said, I have no ideas regarding dynamic parameters, other than making a ton of captures for potentiometers and blending them, depending on the setting (something that NAM developer is currently working on). Maybe you could model some components neurally and some algorhithmically. I am neither a programmer, nor a mathematician, just a gear-obsessed metal musician and a sound engineer who hasn't touched C++ for 7 years, so I can't come up with anything more than that. But all things considered, if you decide to commit to turning LiveSPICE into a more professional product, rather than just a hobby tool, you've got yourself a buyer.

**Comment by edoardogiacomello** [comment]
I think that this is one of the most useful open source projects that targets both electronics and DIY audio.

I'm a data scientist with a PhD in Computer Engineering, but being also a guitarist I like to learn Audio Engineering and building Tube Amplifiers in my free time. To me having this kind of projects as open source is vital, because I don't use them for work but they still allow me to pursue my hobbies.
If they were behind a paywall, I don't think they would be worth the money, no matter how "cool" and functional they might be. 
I spent a lot of money already on guitar modelers thinking they could ease my life as an amateur guitarist, but in the end I went back to real tube amps since they simply fit better the context at reharsals (and surprisingly they are more immediate to use).
It's just an opinion from a random guy, but I think that the same could apply also to this kind of software and could be true for many people.

A professional ECS is meant to be used by companies that can afford much more than a startup software for their core business. If you bring up a paywall for such a useful technology you are basically keeping out a lot of hobbyists that couldn't afford trial and error with real components, just to potentially monetize on producers or musicians that usually don't know much about what's inside the equipment they use.

As flawed this project it might be, I think that it's useful enough to its target audience. I'm literally comparing different revision of a real tube amplifier by plugging my guitar straight in my DAW, it's mindblowing! When it stutters too much I just record a track and render it to hear the differences, but it's still several orders of magnitude faster than simulating with LTSpice alone.

I'd really like to have more spare time to help with the project...

**Comment by TheOtherMrBill** [comment]
I read this thread with great interest. As a guitarist, effects & amp building enthusiast, retired software engineer/architect, I think I _might_ have something to add here. I stumbled upon liveSPICE after dedicating a month to my own vacuum tube amp design software. I have a stack of algorithms that emulate how a tube amplifies a signal and how the different parts of a common gain circuit interact with the mesh of solutions that come from the plate voltage curves. Between that, getting a UI up and running, wrapping a VST around that, I am beyond impressed with the awesomeness that liveSPICE is today. It's 12 years beyond where I am now.

In my experience, performance comes as a cost. But, I also come from a history where 16-bit game development was possible to create Wolfenstein and Doom. Among other things, the math is simplified and the use of lookup tables is common to make it fast. Not promising anything, I am unfamiliar with C#, but I am very familiar with C++ and performance tweaks. So, I may be able to see where things can be optimized or architected in such a way to support circuit design, live VST play through, and the round trip engineering to operate between the two.

In my opinion, this is far beyond a hobby tool. Twelve years of development is not a hobby. That's commitment. It also doesn't have to be slated for commercial success. Open source is here because of people like myself. I didn't like anything else I could find out there, so I was writing my own. Now I've found a project that I can download, compile (after fixing ONE bug), and emulate part of a $100k Dumble! This is awesome and I want it to do more of what I'm after.
